### PR TITLE
Use stream allocator adaptor for hash join table

### PR DESCRIPTION
**Pull request by PointKernel**
Depends on https://github.com/NVIDIA/cuCollections/pull/119

This PR replaces the default hash join allocator with the corresponding `rmm::mr::stream_allocator_adaptor`. It accommodates new `cuco::allocator` APIs that don't take stream as input argument.

### CODE REVIEW

**Comment by PointKernel** [comment]
Current CMake changes are dedicated to CI tests only. Will update the main branch git tag once https://github.com/NVIDIA/cuCollections/pull/119 is merged.

File: cpp/cmake/thirdparty/get_cucollections.cmake
Comment (karthikeyann) [inline cpp/cmake/thirdparty/get_cucollections.cmake:24]: Added "DO NOT MERGE" until dependent PR is checked in and tag is available.

**Review by karthikeyann — CHANGES_REQUESTED** [review]
Looks good. 
Holding this until dependent PR is merged.

File: cpp/cmake/thirdparty/get_cucollections.cmake
Comment (jrhemstad) [inline cpp/cmake/thirdparty/get_cucollections.cmake:24]: @PointKernel can you update this now that the cuco PR is merged?

Comment (PointKernel) [reply cpp/cmake/thirdparty/get_cucollections.cmake]: Done.

File: cpp/cmake/thirdparty/get_cucollections.cmake
Comment (harrism) [inline cpp/cmake/thirdparty/get_cucollections.cmake:24]: If the dependency PR is already merged, shouldn't this GIT_TAG be removed and just depend on a release branch?

File: cpp/src/join/join_common_utils.hpp
Comment (harrism) [inline cpp/src/join/join_common_utils.hpp:25]: Order includes nearest to farthest (top to bottom) and separate libraries with a blank line.

```suggestion
#include <rmm/mr/device/polymorphic_allocator.hpp>

#include <cuco/static_multimap.cuh>
```

**Review by harrism — CHANGES_REQUESTED** [review]

File: cpp/cmake/thirdparty/get_cucollections.cmake
Comment (PointKernel) [reply cpp/cmake/thirdparty/get_cucollections.cmake:24]: cuco doesn't have a release yet thus we choose to fetch a specific commit instead of the main branch (`dev`).

**Comment by jrhemstad** [comment]
@harrism @karthikeyann please re-review.

**Review by karthikeyann — APPROVED** [review]

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**
